File: services/server/src/safe_socket/safe_socket.py
import socket


def send_all(sock: socket.socket, data: bytes) -> None:
    """Sends all bytes in data, looping over sock.send to tolerate short
    writes (a single call to send() is not guaranteed to send everything)."""
    total_sent = 0
    while total_sent < len(data):
        # send() puede devolver 0 sin que la conexión esté cerrada (a
        # diferencia de recv()); si eso pasa, simplemente hay que
        # reintentar. Una conexión realmente rota levanta una excepción
        # (BrokenPipeError/ConnectionResetError), no devuelve 0.
        sent = sock.send(data[total_sent:])
        total_sent += sent


def recv_all(sock: socket.socket, size: int) -> bytes:
    """Receives exactly `size` bytes, looping over sock.recv to tolerate
    short reads (a single call to recv() may return fewer bytes than
    requested)."""
    chunks = []
    remaining = size
    while remaining > 0:
        chunk = sock.recv(remaining)
        if not chunk:
            raise ConnectionError("connection closed while receiving")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"io"
)

// SendAll writes all of data to w, looping over Write to tolerate short
// writes (a single call to Write is not guaranteed to write everything).
func SendAll(w io.Writer, data []byte) error {
	totalSent := 0
	for totalSent < len(data) {
		// Write puede devolver n=0 sin error, sin que eso signifique que
		// la conexión se cerró (a diferencia de un Read que devuelve 0
		// bytes junto con io.EOF); si eso pasa, simplemente hay que
		// reintentar.
		n, err := w.Write(data[totalSent:])
		if err != nil {
			return err
		}
		totalSent += n
	}
	return nil
}

// RecvAll reads exactly size bytes from r, looping over Read to tolerate
// short reads (a single call to Read may return fewer bytes than
// requested, even before reaching EOF).
func RecvAll(r io.Reader, size int) ([]byte, error) {
	buf := make([]byte, size)
	totalRead := 0
	for totalRead < size {
		n, err := r.Read(buf[totalRead:])
		totalRead += n
		if err != nil {
			if err == io.EOF && totalRead == size {
				break
			}
			return nil, err
		}
	}
	return buf, nil
}
